File: app/src/main/java/es/etologic/mahjongscoring2/data/repository/local/daos/GamesDao.java
package es.etologic.mahjongscoring2.data.repository.local.daos;

import android.arch.persistence.room.Dao;
import android.arch.persistence.room.Insert;
import android.arch.persistence.room.OnConflictStrategy;
import android.arch.persistence.room.Query;
import android.arch.persistence.room.Update;
import android.database.sqlite.SQLiteConstraintException;

import java.util.List;

import es.etologic.mahjongscoring2.domain.entities.Game;

@Dao
public interface GamesDao {

    @Insert (onConflict = OnConflictStrategy.ROLLBACK)
    long insertOne(Game game) throws SQLiteConstraintException;
    //TODO: probar con tests. ¿Hace falta la exception o devolvería un 0?

    @Query("SELECT * FROM Games WHERE gameId = :gameId")
    Game getOne(long gameId);

    @Query("SELECT * FROM Games")
    List<Game> getAll();

    @Query("DELETE FROM Games WHERE gameId = :gameId")
    int deleteOne(long gameId);

    @Update
    int updateOne(Game game);
}

File: app/src/main/java/com/etologic/mahjongscoring2/data_source/local_data_source/local/daos/GamesDao.kt
/*
 *     Copyright © 2023  Ernesto Vega de la Iglesia Soria
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package com.etologic.mahjongscoring2.data_source.local_data_source.local.daos

import android.database.sqlite.SQLiteConstraintException
import androidx.room.*
import com.etologic.mahjongscoring2.data_source.model.DBGame
import com.etologic.mahjongscoring2.data_source.model.GameId
import kotlinx.coroutines.flow.Flow

@Dao
interface GamesDao {

    @Query("SELECT * from Games ORDER BY startDate DESC")
    fun getAll(): Flow<List<DBGame>>

    @Query("SELECT * from Games WHERE gameId = :gameId")
    fun getOne(gameId: GameId): Flow<DBGame>

    @Insert(onConflict = OnConflictStrategy.ABORT)
    @Throws(SQLiteConstraintException::class)
    suspend fun insertOne(dbGame: DBGame): GameId

    @Update
    suspend fun updateOne(dbGame: DBGame): Int

    @Query("DELETE FROM Games WHERE gameId = :gameId")
    suspend fun deleteOne(gameId: GameId): Int
}
